### PR TITLE
294 get an error when the number of labels in calibration is not the same as during the training when using cv=prefit

### PR DESCRIPTION
**Pull request by vincentblot28**

### CODE REVIEW

**Review by thibaultcordier — APPROVED** [review]

**Comment by thibaultcordier** [comment]
After my review., this PR is ready.

File: mapie/classification.py [hunk 1]
```diff
@@ -943,15 +952,38 @@ def fit(
...
         self.n_features_in_ = check_n_features_in(X, cv, estimator)
 
         n_samples = _num_samples(y)
-        self.n_classes_ = len(np.unique(y))
-        self.classes_ = np.unique(y)
+
+        if self.cv == "prefit":
+            self.classes_ = estimator.classes_
+            self.n_classes_ = len(np.unique(self.classes_))
+            if self.n_classes_ == len(np.unique(y)):
```
Comment (LacombeLouis) [inline mapie/classification.py:966]: Give a name to the `len(np.unique(y))`, you use it enough and a understandable variable name would give a better insight into what it is.

File: mapie/classification.py [hunk 2]
```diff
@@ -943,15 +952,38 @@ def fit(
 
         estimator = check_estimator_classification(
             X,
-            y_enc,
+            y,
```
Comment (LacombeLouis) [inline mapie/classification.py:955]: Do you ever make a check for the y_enc? By that I mean, do you check that the output of the y_enc is similar or as expected given its size?

File: mapie/classification.py [hunk 3]
```diff
@@ -943,15 +952,38 @@ def fit(
...
+                raise ValueError(
+                    "You have more labels in the y of your calibration dataset"
+                    + " than during training (training"
+                    + f" has {len(np.unique(estimator.classes_))} unique"
+                    + f" labels while calibration have {len(np.unique(y))}"
+                    + " unique labels"
+                )
+
+        else:
+            self.n_classes_ = len(np.unique(y))
```
Comment (LacombeLouis) [inline mapie/classification.py:985]: I think it would make more sense to have these variables be set once we have trained the estimator (in the no-prefit context)? And then have a test later to check the coherence of the results?

File: mapie/classification.py [hunk 4]
```diff
@@ -943,15 +952,38 @@ def fit(
...
 
         n_samples = _num_samples(y)
-        self.n_classes_ = len(np.unique(y))
-        self.classes_ = np.unique(y)
+
+        if self.cv == "prefit":
+            self.classes_ = estimator.classes_
+            self.n_classes_ = len(np.unique(self.classes_))
+            if self.n_classes_ == len(np.unique(y)):
+                pass
```
Comment (LacombeLouis) [inline mapie/classification.py:967]: Is this the best way to implement the tests? Personally, not the biggest fan of the `pass`.

File: mapie/classification.py [hunk 5]
```diff
@@ -943,15 +952,38 @@ def fit(
...
             cv,
             self.estimator
         )
         self.n_features_in_ = check_n_features_in(X, cv, estimator)
 
         n_samples = _num_samples(y)
-        self.n_classes_ = len(np.unique(y))
-        self.classes_ = np.unique(y)
+
+        if self.cv == "prefit":
```
Comment (LacombeLouis) [inline mapie/classification.py:963]: Could we not implement this as a function? say: `check_consistency_labels()`?

File: mapie/tests/test_classification.py [hunk 6]
```diff
@@ -1370,3 +1370,56 @@ def test_error_raps_cv_not_prefit(cv: Union[int, None]):
     mapie = MapieClassifier(method="raps", cv=5)
     with pytest.raises(ValueError, match=r".*RAPS method can only.*"):
         mapie.fit(X_toy, y_toy)
+
+
+def test_not_all_label_in_calib():
+    """Test that the true label cumsumed probabilities
```
Comment (LacombeLouis) [inline mapie/tests/test_classification.py:1376]: I like it better like this:
"""
    Test
"""

File: mapie/tests/test_classification.py [hunk 7]
```diff
@@ -1370,3 +1370,56 @@ def test_error_raps_cv_not_prefit(cv: Union[int, None]):
...
+        cv="prefit"
+    )
+    with pytest.warns(
+        UserWarning, match=r".*WARNING: your calibration dataset.*"
+    ):
+        mapie_clf.fit(X_mapie, y_mapie)
+
+
+def test_calib_have_more_label_raise_error():
+    """Test that the true label cumsumed probabilities
```
Comment (LacombeLouis) [inline mapie/tests/test_classification.py:1410]: See comment about the styling (above) and maybe implement overall?

File: mapie/tests/test_classification.py [hunk 8]
```diff
@@ -1370,3 +1370,56 @@ def test_error_raps_cv_not_prefit(cv: Union[int, None]):
...
+        estimator=clf, method="cumulated_score",
+        cv="prefit"
+    )
+    with pytest.warns(
+        UserWarning, match=r".*WARNING: your calibration dataset.*"
+    ):
+        mapie_clf.fit(X_mapie, y_mapie)
+
+
+def test_calib_have_more_label_raise_error():
```
Comment (LacombeLouis) [inline mapie/tests/test_classification.py:1409]: Missing `--> None`

File: mapie/tests/test_classification.py [hunk 9]
```diff
@@ -1370,3 +1370,56 @@ def test_error_raps_cv_not_prefit(cv: Union[int, None]):
     mapie = MapieClassifier(method="raps", cv=5)
     with pytest.raises(ValueError, match=r".*RAPS method can only.*"):
         mapie.fit(X_toy, y_toy)
+
+
+def test_not_all_label_in_calib():
```
Comment (LacombeLouis) [inline mapie/tests/test_classification.py:1375]: Missing `--> None`

**Review by LacombeLouis — COMMENTED** [review]
Thank you for solving this issue! I think some little details need another review, but good to have checked for this niche scenario!

File: mapie/classification.py (hunk 1 above)
Comment (vincentblot28) [reply mapie/classification.py:966]: done

File: mapie/classification.py (hunk 2 above)
Comment (vincentblot28) [reply mapie/classification.py:955]: I think it is something that must have been tested directly be sklearn non ?

File: mapie/classification.py (hunk 3 above)
Comment (vincentblot28) [reply mapie/classification.py:985]: done

File: mapie/classification.py (hunk 4 above)
Comment (vincentblot28) [reply mapie/classification.py:967]: done

File: mapie/tests/test_classification.py (hunk 9 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py:1375]: done

File: mapie/tests/test_classification.py (hunk 6 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py:1376]: done

File: mapie/tests/test_classification.py [hunk 10]
```diff
@@ -1370,3 +1370,56 @@ def test_error_raps_cv_not_prefit(cv: Union[int, None]):
...
+    """Test that the true label cumsumed probabilities
+    have the correct shape.
+    """
+    clf = LogisticRegression()
+    clf.fit(X, y)
+    X_mapie = X[y != 2]
```
[the remark this answers is not in the document]
Comment (vincentblot28) [reply mapie/tests/test_classification.py:1381]: DOne

File: mapie/tests/test_classification.py (hunk 9 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py:1375]: done

File: mapie/tests/test_classification.py (hunk 7 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py:1410]: done

File: mapie/classification.py (hunk 5 above)
Comment (vincentblot28) [reply mapie/classification.py]: done

File: mapie/tests/test_classification.py [hunk 11]
```diff
@@ -1363,10 +1371,136 @@ def test_get_true_label_position(y_true_proba_place: List[NDArray]):
...
+
+
+def test_not_all_label_in_calib() -> None:
+    """
+    Test that the true label cumsumed probabilities
+    have the correct shape.
+    """
+    clf = LogisticRegression()
+    clf.fit(X, y)
+    indices_remove = y != 2
```
Comment (LacombeLouis) [inline mapie/tests/test_classification.py:1391]: Not the biggest fan of this method to find indices. A numpy function would be better I believe.

File: mapie/classification.py [hunk 12]
```diff
@@ -888,6 +888,63 @@ def _find_lambda_star(
...
+        Raises
+        ------
+        ValueError
+            If number of calibration labels is higher than number of labels
+            for training (in prefit setting)
+        Warning
+            If number of calibration labels is lower than number of labels
+            for training (in prefit setting)
+        """
+        n_unique_y_labels = len(np.unique(y))
```
Comment (LacombeLouis) [inline mapie/classification.py:921]: Do you take into account the use case when we do have the same number of labels but maybe in the training set you have [0, 1, 2] and in the calibration you have [1, 2, 3]? 

I'm not sure this is taken into account.

Comment (LacombeLouis) [reply mapie/classification.py:921]: Or tested for.

**Review by LacombeLouis — COMMENTED** [review]
Still one final small query :)

File: mapie/classification.py (hunk 12 above)
Comment (vincentblot28) [reply mapie/classification.py:921]: Tested in `test_y_1_to_l_minus_1`

Comment (vincentblot28) [reply mapie/classification.py:921]: and in `def test_same_result_y_numeric_and_1_to_l_minus_1`

File: mapie/classification.py [hunk 13]
```diff
@@ -888,6 +888,69 @@ def _find_lambda_star(
...
+        if self.cv == "prefit":
+            classes = estimator.classes_
+            n_classes = len(np.unique(classes))
+            if not set(np.unique(y)).issubset(classes):
+                raise ValueError(
+                    "Values in y do not matched values in estimator.classes_."
```
Comment (LacombeLouis) [inline mapie/classification.py:929]: Forgot the {estimator.classes_}.

Comment (vincentblot28) [reply mapie/classification.py:928]: I'm afraid it would be to heavy if I print all the labels, that's why I didn't used a f-string

File: mapie/classification.py (hunk 2 above)
Comment (LacombeLouis) [reply mapie/classification.py:1000]: Let's hope so. I think generally, we'll need to create a refacto for everything that's directed at label encoding/missing labels.

File: mapie/tests/test_classification.py (hunk 11 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py:1391]: done

File: mapie/tests/test_classification.py [hunk 14]
```diff
@@ -1363,10 +1371,134 @@ def test_get_true_label_position(y_true_proba_place: List[NDArray]):
...
```
Comment (LacombeLouis) [inline mapie/tests/test_classification.py]: Would be best to use `np.where()` I believe.

File: mapie/tests/test_classification.py [hunk 15]
```diff
@@ -1363,10 +1371,134 @@ def test_get_true_label_position(y_true_proba_place: List[NDArray]):
...
+    assert y_pss.shape == (len(X), len(np.unique(y)), 1)
+
+
+def test_warning_not_all_label_in_calib() -> None:
+    """
+    Test that a warning is raised y is binary.
+    """
+    clf = LogisticRegression()
+    clf.fit(X, y)
+    indices_remove = y != 2
```
Comment (LacombeLouis) [inline mapie/tests/test_classification.py]: Would be best to use `np.where()` I believe.

File: mapie/tests/test_classification.py [hunk 16]
```diff
@@ -1363,10 +1371,134 @@ def test_get_true_label_position(y_true_proba_place: List[NDArray]):
...
+
+
+def test_n_classes_prefit() -> None:
+    """
+    Test that the attribute n_classes_ has the correct
+    value with cv="prefit".
+    """
+    clf = LogisticRegression()
+    clf.fit(X, y)
+    indices_remove = y != 2
```
Comment (LacombeLouis) [inline mapie/tests/test_classification.py]: Would be best to use `np.where()` I believe.

**Review by LacombeLouis — COMMENTED** [review]
Some missing`np.where()`.

File: mapie/tests/test_classification.py (hunk 16 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py]: done

File: mapie/tests/test_classification.py (hunk 15 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py]: done

File: mapie/tests/test_classification.py (hunk 14 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py]: done

**Review by LacombeLouis — APPROVED** [review]
This works for me. I think we'll need to make a larger refacto in regard to missing classes. But this is a solid fix/possibly foundation for this larger refacto.

Specially with the fact that we might be making conformal predictions conditional on data soon!

**Review by thibaultcordier — APPROVED** [review]
Agree with previous comments.
